OpenTrons/opentrons-api
# refactor(app,api): output opentrons.cli results directly to a JSON file

## Overview

This PR modifies the `opentrons.cli` analyis command line tool and the app to ensure:

- Protocols that output to stdout (e.g. via `print`) do not corrupt the analysis output
- Expected protocol reading errors are properly communicated to the user

Closes https://github.com/Opentrons/opentrons/issues/10142


## Changelog

- refactor(app-shell): rework the app-shell's analysis logic to increase test coverage and defer analysis file writing to the CLI tool
  - Moved Python path selection: `app-shell/src/protocol-analysis/getPythonPath.ts`
  - Moved opentrons.cli execution: `app-shell/src/protocol-analysis/executeAnalyzeCli.ts`
  - Moved unexpected failure handling: `app-shell/src/protocol-analysis/writeFailedAnalysis.ts`
- refactor(api): rework the opentrons.cli module to output to a JSON file rather than stdout and print expected read errors to stdout

## Review requests

Acceptance test procedures:

1. Open protocol in app
2. Check for success or error message

### Good protocol with `print`

Expected: success

```python
metadata = {
    "apiLevel": "2.12",
}

def run(ctx) -> None:
    print("HELLO WORLD")

    tr = ctx.load_labware("opentrons_96_tiprack_300ul", 1)
    right = ctx.load_instrument("p300_single", "right", [tr])
    right.pick_up_tip()
    right.return_tip()
```

### Bad protocol with missing metadata

Expected: `Error: metadata.apiLevel missing or not statically analyzable in testosaur_v2.py.`

```python
def run(ctx) -> None:
    tr = ctx.load_labware("opentrons_96_tiprack_300ul", 1)
    right = ctx.load_instrument("p300_single", "right", [tr])
    right.pick_up_tip()
    right.return_tip()
```


## Risk assessment

Low, bug fixes
